ADR: CA bundle injection for webhooks (#522)
stackabletech/documentation: modules/contributor/pages/adr/ADR033-foundation-webhooks-ca-bundle.adoc

= ADR033: Foundation for admission or conversion webhooks - CA bundle injection
Doc Writer <doc.writer@asciidoctor.org>
v0.1
:status: draft
:date: 2024-01-09

* Status: {status}
* Deciders:
** Andrew Kenworthy
** Malte Sander
** Sascha Lautenschlaeger
* Date: {date}

Technical Story: https://github.com/stackabletech/issues/issues/361

== Context

There are many use cases for the future development of the SDP that involve validating, mutating and conversion webhooks. Tasks that must be tackled in the near future include:

- The proper versioning of our CRDs and the possibility of smoothly up- and downgrading between versions including breaking changes (see https://github.com/stackabletech/documentation/issues/273)
- Fixing the commons operator initial restarting problem (see https://github.com/stackabletech/commons-operator/issues/111 and spike https://github.com/stackabletech/commons-operator/tree/spike/sts-restarter-webhook)

> When creating an API request, e.g. creating a custom resource, the Kubernetes API server performs the following steps before storing into etcd.

Reviewer: suggestion: Be slightly more specific how the interception works:

```suggestion
When the Kubernetes API server receives an incoming request, for example when a custom resource should be created, webhooks can intercept that request after authentication and authorization. The following steps are performed before the server finally persists the object into etdc.
```

Reviewer: We should be careful to not mix the "admission webhook" concepts with the "CRD versioning webhook" concepts. As far as I know, the versioning webhook lives outside of the flow described below.

Reviewer: If we provide explanations for both concepts, we should clearly separate them into different sections.

Author: In general im with you, but i think in the context of CA injection they behave the same?
Step 5 does not have to include a conversion webhook, kubernetes can just stupidly replace the version itself :D

Reviewer: Yes, in the context of CA injection, both concepts should use the same mechanisms.

1. Authentication / Authorization
2. Mutating admission
3. Schema validation
4. Validation admission
5. Conversion to storage version

> The steps 2 to 4 may run in a loop if multiple mutating admissions are performed and have side effects.

Reviewer: suggestion: There *can* be side effects, but there don't have to be:

```suggestion
The steps 2 to 4 may run in a loop if multiple mutating admissions are performed and can have side effects.
```

    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct
----

=== Validating / Mutating webhooks CA configuration

[source,yaml]
----
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration # or MutatingWebhookConfiguration
webhooks:
  - name: my-webhook.example.com
    matchPolicy: Equivalent
    rules:
      - operations: ['CREATE','UPDATE']
        apiGroups: ['*']
        apiVersions: ['*']
        resources: ['*']
    failurePolicy: "Ignore" # Fail-open (optional)
    sideEffects: None
    clientConfig:
      service:
        namespace: default
        name: example-validation-webhook-server # or example-mutating-webhook-server
        path: /validate # or /mutate
      caBundle: "Ci0tLS0tQk...<base64-encoded PEM bundle>...tLS0K"
----

The `clientConfig.caBundle` can not be shipped by us as it will differ from cluster to cluster and must be injected at runtime.

This ADR is about how to achieve the CA bundle injection using external tools or a self-made solution via the secret-operator.

== Problem Statement

> The required CA bundles for the webhooks endpoints must be injected at runtime. There exist tools like [Cert Manager]([url](https://cert-manager.io/docs/concepts/ca-injector/#injecting-the-kubernetes-api-server-ca)) that do exactly this, adding one of these annotations to an injectionable source: cert-manager.io/inject-ca-from, cert-manager.io/inject-ca-from-secret, or cert-manager.io/inject-apiserver-ca (depending on the injection source). For example:

Reviewer: suggestion: slight reword and listing the different labels:

```suggestion
The required CA bundles for the webhooks endpoints must be injected at runtime. There exist tools like https://cert-manager.io/docs/concepts/ca-injector/#injecting-the-kubernetes-api-server-ca[Cert Manager] that do exactly this.
Adding one of these annotations to an injectionable source:

. `cert-manager.io/inject-ca-from`
. `cert-manager.io/inject-ca-from-secret`
. `cert-manager.io/inject-apiserver-ca`

For example:
```

[source,yaml]
----
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  name: my-webhook.example.com
  annotations:
    cert-manager.io/inject-ca-from-secret: default/my-webhook-example-com-ca

> ----

Reviewer: note: As I stated above, we need to be careful when mixing both concepts. If we provide this example, we would also need to provide an example for the CRD versioning webhook.

Reviewer: Adding another example might be the better way forward. This way we can show that the CA injection should work with both concepts.

Author: I dont follow, what other example do you want? There is one for conversion and validating webhook?

Reviewer: Oh sorry, maybe something like:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.example.com
  annotations:
    cert-manager.io/inject-ca-from-secret: default/my-webhook-example-com-ca
```

Author: ah yes, done https://github.com/stackabletech/documentation/pull/522/commits/256812e2ca3119b74b926f08ebd29c67fe78149a

This can be used for CRDs and conversion webhooks as well.

> The SDP should be compatible with as many of these as possible. The normal way of configuration seems to be adding an annotation to the objects containing the CA, which we can probably somehow support in our Helm chart, or solve via documentation.

Reviewer: note: I'm actually currently working on supporting settings custom labels when deploying our operators via the Helm chart (using the values.yaml file).

Adding support for annotations via the same mechanism is very similar.

= ADR033: Foundation for admission or conversion webhooks - CA bundle injection
Doc Writer <doc.writer@asciidoctor.org>
v0.1
:status: draft
:date: 2024-01-09

* Status: {status}
* Deciders:
** Andrew Kenworthy
** Malte Sander
** Sascha Lautenschlaeger
* Date: {date}

Technical Story: https://github.com/stackabletech/issues/issues/361

== Context

There are many use cases for the future development of the SDP that involve validating, mutating and conversion webhooks. Tasks that must be tackled in the near future include:

- The proper versioning of our CRDs and the possibility of smoothly up- and downgrading between versions including breaking changes (see https://github.com/stackabletech/documentation/issues/273)
- Fixing the commons operator initial restarting problem (see https://github.com/stackabletech/commons-operator/issues/111 and spike https://github.com/stackabletech/commons-operator/tree/spike/sts-restarter-webhook)
- Inject logging / vector sidecar containers (mutating)

> When the Kubernetes API server receives an incoming request, for example when a custom resource should be created, webhooks can intercept that request after authentication and authorization. The following steps are performed before the server finally persists the object into etc.

Reviewer 2: etcd?
```suggestion
When the Kubernetes API server receives an incoming request, for example when a custom resource should be created, webhooks can intercept that request after authentication and authorization. The following steps are performed before the server finally persists the object into etcd.
```

Author: Thanks, was underlined and i just fixed it :D

This ADR is about how to achieve the CA bundle injection using external tools or a self-made solution via the secret-operator.

== Problem Statement

The required CA bundles for the webhooks endpoints must be injected at runtime. There exist tools like https://cert-manager.io/docs/concepts/ca-injector/#injecting-the-kubernetes-api-server-ca[Cert Manager] that do exactly this.
Adding one of these annotations to an injectable source:

. `cert-manager.io/inject-ca-from`
. `cert-manager.io/inject-ca-from-secret`
. `cert-manager.io/inject-apiserver-ca`

For example:

[source,yaml]
----
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  name: my-webhook.example.com
  annotations:
    cert-manager.io/inject-ca-from-secret: default/my-webhook-example-com-ca
----

This can be used for CRDs and conversion webhooks as well.

The SDP should be compatible with as many of these as possible. The normal way of configuration seems to be adding an annotation to the objects containing the CA, which we can probably somehow support in our Helm chart, or solve via documentation.

> In a second step the operator needs to obtain a certificate signed by the CA, for which we could use the k8search backend of the secret operator and point this at a configmap which again, can be filled by cert-manager or the Service CA operator mentioned above, configured via an annotation.

Reviewer 2: I find this a little unclear, but that's maybe because I have only looked at `cert-manager.io/inject-apiserver-ca` in detail, and in that case, there is no second step for us to execute: the annotation is picked up and the bundle is injected a few moments later.

Reviewer: This should be resolved by f5159ce, right?

Reviewer: Stating the cert-manager docs:

> NOTE: In this case you will have to ensure that your webhook is configured to serve a TLS certificate that has been signed by the Kubernetes cluster CA. The disadvantages of this mechanism are that: you will require access to the private key of the Kubernetes cluster CA and you will need to manually rotate the webhook certificate.

If we deploy the webhook automatically (which is the current plan afaik), we need to deal with this. Did we look into that?

Author: That only applies if you want to use the same CA as the Kubernetes API server.

Reviewer: Yes, that's correct. We should then maybe link (see [here](https://github.com/stackabletech/documentation/blob/f5159cec1fc1108365c1f6802eb4eb85cdbbc210/modules/contributor/pages/adr/ADR033-foundation-webhooks-ca-bundle.adoc?plain=1#L133)) to a different section, for example [this](https://cert-manager.io/docs/concepts/ca-injector/#injecting-ca-data-from-a-certificate-resource) or [this](https://cert-manager.io/docs/concepts/ca-injector/#injecting-ca-data-from-a-secret-resource).

- Fixing the commons operator initial restarting problem (see https://github.com/stackabletech/commons-operator/issues/111 and spike https://github.com/stackabletech/commons-operator/tree/spike/sts-restarter-webhook)
- Inject logging / vector sidecar containers (mutating)

When the Kubernetes API server receives an incoming request, for example when a custom resource should be created, webhooks can intercept that request after authentication and authorization. The following steps are performed before the server finally persists the object into etcd.

1. Authentication / Authorization
2. Mutating admission
3. Schema validation
4. Validation admission
5. Conversion to storage version

The steps 2 to 4 may run in a loop if multiple mutating admissions are performed and can have side effects.

In order for the Kubernetes API server to contact webhook controllers, the webhook endpoints have to present a certificate trusted by the Kubernetes API server. This can be configured similarly for conversion and validating / mutating webhooks as follows.

=== Conversion webhooks CA configuration

> [source,yaml]

Reviewer 3: Maybe add a short introductory text explaining what the reader is looking at here. What are the important bits ? Can this be reduced to a relevant snippet?

Author: Added a text, i like the full example with the different versions...maybe metadata / groups etc could be removed but no strong opinion here.
https://github.com/stackabletech/documentation/pull/522/commits/00a89cd2092283328567ee5fa015c8be5f6f0adc

    webhook:
      # conversionReviewVersions indicates what ConversionReview versions are understood/preferred by the webhook.
      # The first version in the list understood by the API server is sent to the webhook.
      # The webhook must respond with a ConversionReview object in the same version it received.
      conversionReviewVersions: ["v1","v1beta1"]
      clientConfig:
        service:
          namespace: default
          name: example-conversion-webhook-server
          path: /crdconvert
        #caBundle: will be injected from a Secret 'default/example-conversion-webhook-ca' with a 'ca.crt' data key
  # either Namespaced or Cluster
  scope: Namespaced
  names:
    # plural name to be used in the URL: /apis/<group>/<version>/<plural>
    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab
    # kind is normally the CamelCased singular type. Your resource manifests use this.
    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct
----

=== Validating / Mutating webhooks CA configuration

> [source,yaml]

Reviewer 3: Same here.

Author: https://github.com/stackabletech/documentation/pull/522/commits/00a89cd2092283328567ee5fa015c8be5f6f0adc

        path: /validate # or /mutate
      #caBundle: will be injected from a Secret 'default/example-conversion-webhook-ca' with a 'ca.crt' data key
----

The `clientConfig.caBundle` cannot be shipped by us as it will differ from cluster to cluster and must be injected at runtime.

This ADR is about how to achieve the CA bundle injection using external tools or a self-made solution via the secret-operator.

== Problem Statement

The required CA bundles for the webhooks endpoints must be injected at runtime. There exist tools like https://cert-manager.io/docs/concepts/ca-injector/#injecting-ca-data-from-a-certificate-resource[Cert Manager] that do exactly this, adding one of these annotations to an injectable source:

. `cert-manager.io/inject-ca-from`
. `cert-manager.io/inject-ca-from-secret`
. `cert-manager.io/inject-apiserver-ca`

For example injecting a CA from a secret using Cert-Manager:

[source,yaml]
----
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  name: my-webhook.example.com
  annotations:
    cert-manager.io/inject-ca-from-secret: default/my-webhook-example-com-ca
----

> This can be used for CRDs and conversion webhooks as well.

Reviewer 3: I don't understand the CRDs part. Maybe a short example would help

Author: I referenced the example already provided. https://github.com/stackabletech/documentation/pull/522/commits/ffdea691ca4c0c08e8fd898ce6759f364ae668a7

The SDP should focus on one solution for CA injection, internally or externally, but strive for compatibility with as many others as possible. The normal way of configuration seems to be adding an annotation to the objects containing the CA, which can be supported in our Helm chart or solved via documentation.

For clusters without any existing manager we should provide our own, lightweight caBundle injector that can work in tandem with the secret operator.

== Decision Drivers

> * Generic solution to be compatible with as many external cert providers as possible

Reviewer 3: What does it mean to be compatible with others ? At what level ? Reuse labels ? Crypto algos ?

Author: tried to clarify https://github.com/stackabletech/documentation/pull/522/commits/d061e00517c4107f3409347542d28fafe060c501

> * Openshift compatible

Reviewer 3: I forgot to mention that our current CSI based  solution is not OLM compatible. For example, the secret op is a pain to install and impossible to maintain (upgrade).

If we insist in building our own tools, it would be good to have an alternative.

Also, OpenShift now has [it's own secret op](https://docs.openshift.com/container-platform/4.14/storage/container_storage_interface/ephemeral-storage-shared-resource-csi-driver-operator.html)

> * How to activate / deactivate if e.g. no conversion webhooks should be applied?

Reviewer 3: Activate/deactivate bundle injection? What does this have to do with the question if webhooks should be applied or not ?

Author: The question here is about how to inject. With a solution via cert-manager you will add an annotation for the ca injection. For us this is probably done via helm templating.
If customers do not want any conversion (e.g. always use the storage version and update everything accordingly regarding breaking changes) there must be an option in the templating to not provide the webhook at all (therefore do not add the ca injection annotaiton) as well as not deploying the webhook at all.

[[option3]]
=== Lightweight self-made solution via secret-operator

This would be the Stackable internal solution to avoid any external party tools. It would work in a similar way to the OpenShift Service CA operator but would rather inject the bundles via the CSI instead of ConfigMap mounts.

[[option4]]
=== Common library for cert management / injection in operator-rs

Put the cert management / injection stuff into a library in operator-rs (similar to the code in secret-operator) that we run in-process for each operator that has a webhook. See https://github.com/stackabletech/commons-operator/tree/spike/sts-restarter-webhook[commons-operator spike].

== Pros and Cons of the Options

=== <<option1>>

* Good, because covers both Kubernetes and Openshift
* Good, because widely used and de-facto standard for cert injection
* Good, because cert injection works via annotations (compatibility) for various Kubernetes Resources (CRDs, Validating/Mutating webhooks)

> * Bad, because another tool we have to be experienced with, check for updates and breaking features etc.

Reviewer 3: Another major advantage I see is that it's one less (critical) piece of tech that we need to maintain.

Author: Yeah, this is not quite consistent. The disadvantage of that is mentioned ("Bad, because another tool we have to be experienced with, check for updates and breaking features 
etc.")...

Reviewer 3: So here is my suggestion:
```suggestion
* Good, because it's already popular and familiar to users
* Good, because it frees maintenance and development resources for us
* Bad, because another tool we have to be experienced with, check for updates and breaking features etc.
```
And I would delete line 200.

=== <<option2>>

* Good, because cert injection works via annotations (compatibility)
* Bad, because specific to OpenShift
* Bad, because only injecting to ConfigMaps
* Bad, because another tool we have to be experienced with, check for updates and breaking features etc.

=== <<option3>>

* Good, because no external tools are required
* Good, because reusing secret-operator and cert management should be an internal part of the SDP
* Bad, because requires more time and coding
* Bad, because secret-operator must version its own CRDs as well which could be a predicament

=== <<option4>>

* Good, because no external tools are required

> * Good, because all included within the SDP

Reviewer 3: Aren't these two basically saying the same thing? :)

Author: i would say yes :)

Reviewer 3: ```suggestion
```

1. Authentication / Authorization
2. Mutating admission
3. Schema validation
4. Validation admission
5. Conversion to storage version

The steps 2 to 4 may run in a loop if multiple mutating admissions are performed and can have side effects.

In order for the Kubernetes API server to contact webhook controllers, the webhook endpoints have to present a certificate trusted by the Kubernetes API server. This can be configured similarly for conversion and validating / mutating webhooks as follows.

=== Conversion webhooks CA configuration

[source,yaml]
----
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.example.com
  annotations:
    cert-manager.io/inject-ca-from-secret: default/example-conversion-webhook-ca
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: example.com
  # list of versions supported by this CustomResourceDefinition
  versions:
  - name: v1beta1
    # Each version can be enabled/disabled by Served flag.
    served: true
    # One and only one version must be marked as the storage version.
    storage: true

>     # Each version can define its own schema when there is no top-level

Author: This does not really make sense to me :) ?

>     # Each version can define its own schema when there is no top-level

Reviewer 2: the rest of the sentence is on the next line

>     # Each version can define its own schema when there is no top-level

Reviewer 2: so: "when no top-level schema is defined"

>     # Each version can define its own schema when there is no top-level

Reviewer 2: or: "when there is no top-level schema defined"

[[option2]]
=== https://github.com/openshift/service-ca-operator/blob/master/README.md[OpenShift Service CA operator]

The OpenShift Service CA operator is an OpenShift ClusterOperator and contains several controllers:

* *Serving cert signer*: Issues a signed serving certificate/key pair to services annotated with `service.beta.openshift.io/serving-cert-secret-name` via a secret
* *ConfigMap CA bundle injector*: Watches for configmaps annotated with `service.beta.openshift.io/inject-cabundle=true` and adds or updates a data item (key `service-ca.crt`) containing the PEM-encoded CA signing bundle. Consumers of the configmap can then trust `service-ca.crt` in their TLS client configuration, allowing connections to services that utilize service-serving certificates. Pods referencing the `service-ca.crt` in a VolumeMount will not start before the CA bundle was injected.

[[option3]]
=== Lightweight self-made solution via secret-operator

> This would be the Stackable internal solution to avoid any external party tools. It would work similar to the OpenShift Service CA operator but rather injecting the bundles via the CSI instead of ConfigMap mounts.

Reviewer 2: ```suggestion
This would be the Stackable internal solution to avoid any external party tools. It would work in a similar way to the OpenShift Service CA operator but would rather inject the bundles via the CSI instead of ConfigMap mounts.
```

=== <<option1>>

* Good, because covers both Kubernetes and Openshift
* Good, because widely used and de-facto standard for cert injection
* Good, because cert injection works via annotations (compatibility) for various Kubernetes Resources (CRDs, Validating/Mutating webhooks)
* Bad, because another tool we have to be experienced with, check for updates and breaking features etc.

=== <<option2>>

* Good, because cert injection works via annotations (compatibility)
* Bad, because specific to OpenShift
* Bad, because only injecting to ConfigMaps
* Bad, because another tool we have to be experienced with, check for updates and breaking features etc.

=== <<option3>>

* Good, because no external tools are required
* Good, because reusing secret-operator and cert management should be an internal part of the SDP
* Bad, because requires more time and coding
* Bad, because secret-operator must version its own CRDs as well which could be a predicament

=== <<option4>>

* Good, because no external tools are required
* Good, because all included within the SDP
* Good, because external tools can be used if required
* Bad, because requires more time and coding

== Decision Outcome

> Chosen option <<option4>>, because the CA injection will be handled as part of the SDP and no external dependencies are required. The common library will reside in the operator-rs and used in every operator. OpenShift should not pose a problem since no extra components are necessary. The required infrastructure (e.g. operator-templating, ca injection / generation) can be developed in parallel to the operators e.g. version conversion logic.

Reviewer 2: ```suggestion
Chosen option <<option4>>, because the CA injection will be handled as part of the SDP and no external dependencies are required. The common library will reside in the operator-rs and be used in every operator. OpenShift should not pose a problem since no extra components are necessary. The required infrastructure (e.g. operator-templating, ca injection / generation) can be developed in parallel to the operators e.g. version conversion logic.
```

The OpenShift Service CA operator is an OpenShift ClusterOperator and contains several controllers:

* *Serving cert signer*: Issues a signed serving certificate/key pair to services annotated with `service.beta.openshift.io/serving-cert-secret-name` via a secret
* *ConfigMap CA bundle injector*: Watches for configmaps annotated with `service.beta.openshift.io/inject-cabundle=true` and adds or updates a data item (key `service-ca.crt`) containing the PEM-encoded CA signing bundle. Consumers of the configmap can then trust `service-ca.crt` in their TLS client configuration, allowing connections to services that utilize service-serving certificates. Pods referencing the `service-ca.crt` in a VolumeMount will not start before the CA bundle was injected.

[[option3]]
=== Lightweight self-made solution via secret-operator

This would be the Stackable internal solution to avoid any external party tools. It would work in a similar way to the OpenShift Service CA operator but would rather inject the bundles via the CSI instead of ConfigMap mounts.

[[option4]]
=== Common library for cert management / injection in operator-rs

> Put the cert management / injection stuff into a library in operator-rs (similar to the code in secret-operator) that we run in-process for each operator that has a webhook. See https://github.com/stackabletech/commons-operator/tree/spike/sts-restarter-webhook[commons-operator spike].

Reviewer 4: Given this is the chosen option, I think it would be great if you could add one or two additional sentences.
I had a look at the spike branch, but it has 1232 additions and is not self-explaining

Reviewer: I created this PR: https://github.com/stackabletech/operator-rs/pull/736 It will implement the auto-generation of TLS certificates. It will of course add a slew of doc comments about the technical details.

Reviewer: I should be able to refine this section in the ADR once I have some code in place.

Reviewer: This should be addressed in 3277224.
